src/driver-sinowealth.c

```diff
@@ -0,0 +1,475 @@
+	}
+
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		ratbag_resolution_set_dpi_list(resolution, dpis, num_dpis);
+		ratbag_resolution_set_cap(resolution, RATBAG_RESOLUTION_CAP_SEPARATE_XY_RESOLUTION);
+	}
+
+	/* Set up LED capabilities */
+	led = ratbag_profile_get_led(profile, 0);
+	led->type = RATBAG_LED_TYPE_SIDE;
+	led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+	ratbag_led_set_mode_capability(led, RATBAG_LED_OFF);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_ON);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_CYCLE);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_BREATHING);
+	ratbag_led_unref(led);
+
+	/* Set up DPI indicator LEDs */
+	for (int i = 1; i < SINOWEALTH_NUM_DPIS + 1; i++) {
+		led = ratbag_profile_get_led(profile, i);
+		led->type = RATBAG_LED_TYPE_DPI;
+		led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+		led->mode = RATBAG_LED_ON;
+		ratbag_led_set_mode_capability(led, RATBAG_LED_ON);
+		ratbag_led_unref(led);
+	}
+
+	ratbag_profile_unref(profile);
+}
+
+static int
+sinowealth_test_hidraw(struct ratbag_device *device)
+{
+	/* Only the keyboard interface has this report */
+	return ratbag_hidraw_has_report(device, SINOWEALTH_REPORT_ID_CONFIG);
+}
+
+static int
+sinowealth_probe(struct ratbag_device *device)
+{
+	int rc;
+	struct sinowealth_data *drv_data = 0;
+	struct ratbag_profile *profile = 0;
+
+	rc = ratbag_find_hidraw(device, sinowealth_test_hidraw);
+	if (rc)
+		goto err;
+
+	drv_data = zalloc(sizeof(*drv_data));
+	ratbag_set_drv_data(device, drv_data);
+
+	sinowealth_init_profile(device);
+
+	profile = ratbag_device_get_profile(device, 0);
+	rc = sinowealth_read_profile(profile);
+	if (rc) {
+		rc = -ENODEV;
+		goto err;
+	}
+
+	return 0;
+
+err:
+	free(drv_data);
+	ratbag_set_drv_data(device, 0);
+	return rc;
+}
+
+static int
+sinowealth_commit(struct ratbag_device *device)
+{
+	struct ratbag_profile *profile = ratbag_device_get_profile(device, 0);
+	struct sinowealth_data *drv_data = device->drv_data;
+	struct sinowealth_config_report *config = &drv_data->config;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	int rc;
+
+	config->config &= ~SINOWEALTH_XY_INDEPENDENT;
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (resolution->dpi_x != resolution->dpi_y && resolution->dpi_x && resolution->dpi_y) {
+			config->config &= SINOWEALTH_XY_INDEPENDENT;
+			break;
+		}
+	}
+
+	config->dpi_enabled = 0xFF;
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (config->config & SINOWEALTH_XY_INDEPENDENT) {
+			config->dpi[resolution->index * 2] = sinowealth_dpi_to_raw(resolution->dpi_x);
+			config->dpi[resolution->index * 2 + 1] = sinowealth_dpi_to_raw(resolution->dpi_y);
+		} else {
+			config->dpi[resolution->index] = sinowealth_dpi_to_raw(resolution->dpi_x);
+		}
+		if (resolution->dpi_x && resolution->dpi_y) {
+			/* enable DPI step (dpi_enabled is inverted) */
+			config->dpi_enabled &= ~(1<<resolution->index);
```

Comment:
spaces around `<<` for consistency

Reply:
I think it may be easier to understand if you OR the bits first and then invert the lot after the loop

Reply:
Agreed